feat(VotesERC20LockableV1): add `maxTotalSupply` to cap mintable token supply

twodam opened the pull request:
### Summary

* Added `MaxTotalSupplyUpdated` event, `ExceedMaxTotalSupply` error, and a `maxTotalSupply` getter to the `IMintableV1` interface.
* Updated the `VotesERC20LockableV1` contract:
    * Introduced a `maxTotalSupply` variable to cap the total supply during minting.
    * Integrated `_maxTotalSupply` into the `initialize` function.
    * Enforced the `maxTotalSupply` limit within the `mint` function.
* Extended the `VotesERC20LockableV1.test.ts` test suite with:
    * Test cases for the new `setMaxTotalSupply` function.
    * Validations for the `mint` function to ensure supply limits are respected.

twodam committed:
feat(VotesERCLockableV1): add maxTotalSupply variable and setter function which limits the amount that mint function can use

twodam committed:
build(hardhat.config.ts): add viaIR option so it can compile

adamgall commented on hardhat.config.ts:
```diff
@@ -15,6 +15,7 @@ const config: HardhatUserConfig = {
       {
         version: '0.8.30',
         settings: {
+          viaIR: true,
```
I don't want this line added. Let's figure out how to get the contracts to compile without it please.

adamgall replied:
I've fixed this

adamgall committed:
refactor(VotesERC20): streamline allocation handling in initialization

This commit refactors the initialization process of the VotesERC20 contracts by replacing separate allocation addresses and amounts arrays with a single array of Allocation structs. This change simplifies the code and enhances clarity in how initial allocations are managed. Additionally, the hardhat configuration has been updated to remove the viaIR option, ensuring compatibility with the latest Solidity compiler settings.

adamgall commented on contracts/deployables/erc20/VotesERC20LockableV1.sol:
```diff
@@ -77,7 +73,19 @@ contract VotesERC20LockableV1 is ILockableV1, IMintableV1, VotesERC20V1 {
         }
     }
 
+    function setMaxTotalSupply(uint256 newMaxTotalSupply) external onlyOwner {
+        uint256 currentlyMaxTotalSupply = maxTotalSupply;
+        maxTotalSupply = newMaxTotalSupply;
+        if (currentlyMaxTotalSupply != newMaxTotalSupply) {
+            emit MaxTotalSupplyUpdated(newMaxTotalSupply);
+        }
+    }
+
     function mint(address to, uint256 amount) external onlyOwner {
+        uint256 newTotalSupply = totalSupply() + amount;
+        if (newTotalSupply > maxTotalSupply) {
+            revert ExceedMaxTotalSupply();
+        }
         _mint(to, amount);
     }
```
Just to be clear, seems like this code allows for setting a new `maxTotalSupply` which is _less than_ the current amount of minted tokens. It would not allow any MORE tokens to be minted, but it does create a strange state.

Is this what we want?

I'm not even sure how we would be able to check the current amount of minted tokens (in order to provide a safety check)...

adamgall replied:
Wait duh, there's a `totalSupply()` function. We could use that to make sure the `newMaxTotalSupply` doesn't exceed current `totalSupply()`.

twodam replied:
yeah, it didn't check if new `maxTotalSupply` is smaller, I thought it may make sense with future burnable feature so that `totalSupply` can decrease. also we can always call another `setMaxTotalSupply` to adjust it, nothing stuck.

twodam replied:
though you are right, this create a strange state, will update.

twodam committed:
feat(VotesERC20LockableV1): dont allow to setMaxTotalSupply less than totalSupply

adamgall committed:
Merge branch 'develop' into feature/eng-934-make-erc20-mintable-by-owner-with-a-max-total-supply

adamgall commented on contracts/deployables/erc20/VotesERC20LockableV1.sol:
```diff
@@ -77,10 +77,27 @@ contract VotesERC20LockableV1 is IVotesERC20LockableV1, VotesERC20V1 {
+        if (newMaxTotalSupply < totalSupply()) {
+            revert InvalidMaxTotalSupply();
+        }
+        uint256 currentlyMaxTotalSupply = _maxTotalSupply;
+        _maxTotalSupply = newMaxTotalSupply;
+        if (currentlyMaxTotalSupply != newMaxTotalSupply) {
+            emit MaxTotalSupplyUpdated(newMaxTotalSupply);
+        }
```
Nitpick, but why only emit the event if the new value is different than the current value? If you're doing that, I would suggest to include the assignment inside the condition too.

But my preference would be to just remove the condition entirely. Always do the assignment and always emit the event.

twodam replied:
I was trying to match the style used in the `whitelist` function above:
```solidity
function whitelist(address account, bool isWhitelisted) external onlyOwner {
        bool currentlyWhitelisted = whitelisted[account];
        whitelisted[account] = isWhitelisted;
        if (currentlyWhitelisted != isWhitelisted) {
            emit Whitelisted(account, isWhitelisted);
        }
    }
```

adamgall replied:
Ah I see, that makes sense. In retrospect I think I'd like both of those snippets to be updated to the style in my first comment then 😅

twodam replied:
I’ve updated `setMaxTotalSupply`. Maybe you can update `whitelist` separately outside this PR? I believe you have some tests where you prefer that event not to be emitted:
```solidity
describe('When adding the same address again', () => {
  beforeEach(async () => {
    addTx = await proxy.connect(owner).whitelist(whitelistMember.address, true);
  });

  it('should not emit an event', async () => {
    await expect(addTx).to.not.emit(proxy, 'Whitelisted');
  });
});
```

adamgall replied:
Will do!

twodam committed:
feat(VotesERC20LockableV1): always emit event along with assignment

twodam commented:
@adamgall could you merge this PR? I don't have the access to do it.

adamgall commented:
I'll do it tomorrow morning (my time), I'll also try to diagnose why you can't merge...

tbwebb22 reviewed (approved):
LGTM
